conda_forge_feedstock_ops/container_utils.py

```diff
@@ -107,13 +118,148 @@ def _get_proxy_mode_container_args():
+            raise ValueError(
+                f"container_path must be a subdirectory of {CF_FEEDSTOCK_OPS_DIR}"
+            )
+
+    @property
+    def relative_container_path(self) -> PurePosixPath:
+        """
+        The relative path in the container.
+        """
+        # This should never fail because we check in __post_init__
+        # that the container_path is a subdirectory of CF_FEEDSTOCK_OPS_DIR
+        return self.container_path.relative_to(CF_FEEDSTOCK_OPS_DIR)
+
+    @classmethod
+    def to_cf_feedstock_ops_dir(cls, host_path: Path, read_only: bool = True) -> Self:
+        """
+        Creates a new mount to the default CF_FEEDSTOCK_OPS_DIR.
+        """
+        return cls(
+            host_path=host_path,
+            container_path=CF_FEEDSTOCK_OPS_DIR,
+            read_only=read_only,
+        )
+
+
+@contextmanager
+def _mounts_to_tar(mounts: Iterable[VirtualMount]) -> Iterator[IO[bytes]]:
+    """
+    Yields a temporary file with the host path contents of the mounts tarred.
+    """
+    with tempfile.TemporaryFile(mode="wb+", suffix=".tar") as target:
+        with tarfile.open(fileobj=target, mode="w") as tar:
+            for mount in mounts:
+                tar.add(mount.host_path, arcname=mount.relative_container_path)
+        target.flush()
+        target.seek(0)
+        yield target
+
+
+def _delete_non_ignore_paths(target_dir_or_file: Path):
+    """
+    Delete any files and directories in the target directory or file that are
+    NOT in the ignore paths.
+    """
+    for file_path in target_dir_or_file.rglob("*"):
+        relative_path = file_path.relative_to(target_dir_or_file)
+
+        if not VirtualMount.IGNORE_PATHS.intersection(relative_path.parts):
+            if file_path.is_dir():
+                shutil.rmtree(file_path)
+            else:
+                file_path.unlink()
+
+
+def _untar_directory_or_file(
+    tar: TarFile, path_inside_tar: PurePosixPath, target_dir_or_file: Path
+):
+    """
+    Untar a directory or file from the tar file to the target directory.
+    Any existing files and directories are removed, except if they were skipped
+    because they are in the ignore paths.
+    """
+    _delete_non_ignore_paths(target_dir_or_file)
+
+    members = (
+        m
+        for m in tar.getmembers()
+        if m.name.startswith(str(path_inside_tar) + "/")
+        or m.name == str(path_inside_tar)
+    )
+    members = (
+        m
+        for m in members
+        if not any(part in VirtualMount.IGNORE_PATHS for part in m.name.split("/"))
+    )
+
+    # note that filter="data" is crucial to prevent security issues - the tar file
+    # is untrusted!
+    for member in members:
+        # set_attrs keeps permissions in a safe way
+        # https://docs.python.org/3/library/tarfile.html
+        tar.extract(member, target_dir_or_file.parent, set_attrs=True, filter="data")
+
+
+def _untar_mounts_from_stream(
```

Comment:
Note to myself: This should be public as long as we use a custom image in cf-scripts